### PR TITLE
Isly

### PR DESCRIPTION
**Pull request by EliasEriksson**
* converted non permission interfaces to use isly
  larger reworks are expected for permissions no time was put down to convert those
* updated imports

### CODE REVIEW

**Review by EliasEriksson — COMMENTED** [review]
self check.
userwidgets/api builds with no errors/warnings after these changes

File: Organization/Creatable.ts [hunk 1]
```diff
@@ -1,32 +1,31 @@
...
+import { Email } from "../Email"
 import { Permissions } from "../User/Permissions"
-
+import { Identifier } from "./Identifier"
 export interface Creatable {
-	id?: string
+	id?: Identifier
 	name: string
 	permissions: string[]
 	users: { email: string; permissions?: [Permissions.Application, Permissions.Organization] }[]
```
Comment (TheLillyleaf) [inline Organization/Creatable.ts:9]: email:Email

File: Organization/index.spec.ts [hunk 2]
```diff
@@ -1,16 +1,17 @@
-import * as isoly from "isoly"
-import * as model from "../index"
+import { isoly } from "isoly"
+import { userwidgets } from "../index"
+
 describe("Organization", () => {
 	const now = isoly.DateTime.now()
-	const creatable: model.Organization = {
-		id: "myId",
+	const creatable: userwidgets.Organization = {
```
Comment (TheLillyleaf) [inline Organization/index.spec.ts:6]: Är ingen creatable, borde heta annat.

File: Organization/index.ts [hunk 3]
```diff
@@ -1,27 +1,29 @@
...
-import { Creatable as CreatableOrganization } from "./Creatable"
+import { isoly } from "isoly"
+import { isly } from "isly"
+import { Email } from "../Email"
+import { Creatable as OrganizationCreatable } from "./Creatable"
+import { Identifier as OrganizationIdentifier } from "./Identifier"
 
-export interface Organization extends Omit<CreatableOrganization, "users"> {
-	id: string
+export interface Organization extends Omit<OrganizationCreatable, "users"> {
```
Comment (TheLillyleaf) [inline Organization/index.ts:7]: Organization.Creatable

File: User/Credentials/index.ts [hunk 4]
```diff
@@ -1,17 +1,21 @@
 import * as authly from "authly"
```
Comment (TheLillyleaf) [inline User/Credentials/index.ts:1]: Ska vi fixa?

File: User/Feedback/Invitation/index.ts [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
+import { gracely } from "gracely"
+import { isly } from "isly"
 import { Invitation as InvitationInterface } from "./Invitation"
 
 export type Invitation = InvitationInterface | gracely.Error
 export namespace Invitation {
-	export function is(value: Invitation | any): value is Invitation {
-		return InvitationInterface.is(value) || gracely.Error.is(value)
-	}
+	export const type = isly.union<Invitation, InvitationInterface, gracely.Error>(
```
Comment (TheLillyleaf) [inline User/Feedback/Invitation/index.ts:7]: Invitation.Interface

File: User/Feedback/Notification/index.ts [hunk 6]
```diff
@@ -1,12 +1,16 @@
-import * as gracely from "gracely"
+import { gracely } from "gracely"
+import { isly } from "isly"
 import { Notification as NotificationInterface } from "./Notification"
 
 export type Notification = NotificationInterface | gracely.Error
```
Comment (TheLillyleaf) [inline User/Feedback/Notification/index.ts:5]: Notification.Interface

File: User/Invite/index.ts [hunk 7]
```diff
@@ -1,8 +1,9 @@
 import * as isoly from "isoly"
```
Comment (TheLillyleaf) [inline User/Invite/index.ts:1]: New update

File: User/Invite/index.ts [hunk 8]
```diff
@@ -1,8 +1,9 @@
...
 import * as authly from "authly"
-import { Creatable as CreatableInvite } from "./Creatable"
+import { isly } from "isly"
+import { Creatable as InviteCreatable } from "./Creatable"
 
-export interface Invite extends CreatableInvite {
+export interface Invite extends Invite.Creatable {
 	issuer: string
```
Comment (TheLillyleaf) [inline User/Invite/index.ts:7]: Isn't this an email?

File: User/Invite/index.ts [hunk 9]
```diff
@@ -34,18 +35,19 @@ const transformers: authly.Property.Transformer[] = [
...
-			CreatableInvite.is(value) &&
-			typeof value.issuer == "string" &&
-			typeof value.audience == "string" &&
-			isoly.DateTime.is(value.issued) &&
-			isoly.DateTime.is(value.expires) &&
-			typeof value.token == "string"
-		)
-	}
+	export const type = Creatable.type.extend<Invite>({
+		issuer: isly.string(/.+/),
```
Comment (TheLillyleaf) [inline User/Invite/index.ts:43]: If emai, change here.

File: User/Key/index.ts [hunk 10]
```diff
@@ -1,8 +1,9 @@
-import * as isoly from "isoly"
+import { isoly } from "isoly"
 import * as authly from "authly"
-import { Creatable as CreatableKey } from "./Creatable"
+import { isly } from "isly"
+import { Creatable as KeyCreatable } from "./Creatable"
 
-export interface Key extends CreatableKey {
+export interface Key extends Key.Creatable {
 	issuer: string
```
Comment (TheLillyleaf) [inline User/Key/index.ts:7]: if Email, change here and isly

File: User/Permissions/Application.ts [hunk 11]
```diff
@@ -8,7 +8,7 @@ export interface Application {
 	[resource: string]: Permission | undefined
 }
 export namespace Application {
-	export function is(value: Application | any): value is Application & Record<string, any> {
```
Comment (TheLillyleaf) [inline User/Permissions/Application.ts:11]: Permissions är inte isly, finns det någon anledning till det ?

File: User/index.ts [hunk 12]
```diff
@@ -9,24 +11,24 @@ import { Password as UserPassword } from "./Password"
...
+	export const type = isly.object<User>({
+		email: Email.type,
+		name: Name.type,
+		permissions: isly.fromIs("Permissions", UserPermissions.is),
+		created: isly.fromIs("isoly.DateTime", isoly.DateTime.is),
+		modified: isly.fromIs("isoly.DateTime", isoly.DateTime.is),
+	})
+	export const is = type.is
+	export const flaw = type.flaw
 	export function toKey(user: User, applicationId: string, organizationIds?: string[]): UserKey.Creatable | undefined {
```
Comment (TheLillyleaf) [inline User/index.ts:32]: application["id] och organization["id"][]

**Review by TheLillyleaf — CHANGES_REQUESTED** [review]
Some small thuings

File: User/Invite/index.ts (hunk 8 above)
Comment (EliasEriksson) [reply User/Invite/index.ts:7]: it is the name of the authentication server who issued the JWT

File: User/Permissions/Application.ts (hunk 11 above)
Comment (EliasEriksson) [reply User/Permissions/Application.ts:11]: kommer förmodligen skrivas om senare denna vecka.

**Review by TheLillyleaf — APPROVED** [review]

**Pull request merged**